ocaml-lsp-server/src/compl.ml

```diff
@@ -134,62 +130,152 @@ let range_prefix (lsp_position : Position.t) prefix : Range.t =
+      Query_commands.dispatch pipeline command
+    with
+    | r -> Ok r
+    | exception e -> Error (Exn_with_backtrace.capture e)
+
+  let process_dispatch_resp = function
+    | Error { Exn_with_backtrace.exn = Merlin_analysis.Construct.Not_a_hole; _ }
+      ->
+      []
+    | Error e -> Exn_with_backtrace.reraise e
+    | Ok (loc, constructed_exprs) ->
+      let range = Range.of_loc loc in
+      let completionItem_of_constructed_expr idx expr =
+        let textEdit = `TextEdit { TextEdit.range; newText = expr } in
+        let command =
+          Client.Vscode.Commands.Custom.next_hole ~start_position:range.start
+            ~notify_if_no_hole:false ()
+        in
+        CompletionItem.create ~label:expr ~textEdit ~filterText:("_" ^ expr)
+          ~kind:CompletionItemKind.Text ~sortText:(sortText_of_index idx)
+          ~command ()
+      in
+      List.mapi constructed_exprs ~f:completionItem_of_constructed_expr
+end
+
+let completion_list ?(isIncomplete = false) items =
+  `CompletionList (CompletionList.create ~isIncomplete ~items)
+
+let complete doc pos =
+  let position = Position.logical pos in
+  let prefix =
+    prefix_of_position ~short_path:false (Document.source doc) position
   in
-  let textDocument = TextDocumentIdentifier.create ~uri:(Document.uri doc) in
-  let compl_info =
-    CompletionParams.create ~textDocument ~position:lsp_position ()
-    |> CompletionParams.yojson_of_t
+  let open Fiber.O in
+  let complete_prefix_list () =
+    let+ items = Complete_by_prefix.complete doc prefix pos in
+    completion_list items
   in
-  let items = List.mapi ~f:(item ~compl_info) items in
-  `CompletionList { CompletionList.isIncomplete = false; items }
+  if not (Typed_hole.can_be_hole prefix) then
+    complete_prefix_list ()
+  else
+    let reindex_sortText completion_items =
+      List.mapi completion_items ~f:(fun idx (ci : CompletionItem.t) ->
+          let sortText = Some (sortText_of_index idx) in
+          { ci with sortText })
+    in
+    let preselect_first = function
+      | [] -> []
+      | ci :: rest -> { ci with CompletionItem.preselect = Some true } :: rest
+    in
+    (* TODO: do merlin construct and complete_prefix dispatches at once *)
+    let+ construct_cmd_resp, compl_by_prefix_resp =
+      Document.with_pipeline_exn doc (fun pipeline ->
+          let construct_cmd_resp =
+            Complete_with_construct.dispatch_cmd position pipeline
+          in
+          let compl_by_prefix_resp =
+            Complete_by_prefix.dispatch_cmd ~prefix position pipeline
+          in
+          (construct_cmd_resp, compl_by_prefix_resp))
+    in
+    let construct_completionItems =
+      Complete_with_construct.process_dispatch_resp construct_cmd_resp
+    in
+    let compl_by_prefix_completionItems =
+      match compl_by_prefix_resp with
+      | Ok resp -> Complete_by_prefix.process_dispatch_resp doc pos resp
+      | Error _ ->
+        (* TODO: log the error in future *)
```

Comment:
All of this `dispatch_cmd` vs `dispatch_cmd_exn` stuff was introduced to make it possible to recover hole suggestions if completion fails, right? That doesn't like a good trade-off to me. There's no need to try and recover from bugs in the server, let's just fail the request and move on. The important thing here is that the user can give us a useful bug report, and not if some partial functionality is preserved.

Reply:
If your main argument is to avoid failing silently: 

I agree that failing silently is undesirable, but I think showing a warning message there may be slightly more appropriate there:

1. If the completion by prefix fails at that particular place, it's very likely because of a merlin problem, not server. 
2. If a user doesn't get completion suggestions once in a while (since completion bugs are quite rare from my experience), they go on with their lives, but if they get a warning message prompting to submit a bug report with an appropriate log with the backtrace from the server, they are much more likely to report things. 

So I think it may be better here to show a warning, log the exception, and offer our best completion suggestions we could deliver. Wdyt? 

If you more want to avoid the redundancy of `dispatch_cmd` and `dispatch_cmd_exn`, I also agree. I could leave only non-exn one and adapt `process_dispatch_resp` to work with a result, but it would just be reraising.

Side notes: 
- I'm really sorry for long argumentation. I really have the best in mind for the user.
- We miss a good logger in ocaml-lsp. In my understanding, if we put the logger in `State.t`, we will have to propagate it as an argument across all functions, say in this PR: `complete -> dispatch_cmd -> ...`. Should we maybe have a global logger (in `lsp-fiber` library?) that keeps an option ref, initialized by ocaml-lsp-server?

Reply:
The issue that I'm pointing out isn't that we're failing silently, but that we're trying too hard to recover from buggy code. The instant we know we're in the buggy path, we should just abort everything and do our best to guide the user to submit a bug report. Of course we can't take this literally and  terminate the server in such instances - it would make the user experience very bad, but this particular edge case doesn't deserve any special treatment.

The code inside the dispatch should just get the completion and typed hole candidates. Should this fail for any reason, then it's OK to just abort the entire completion request. The user will get an appropriate error message in their client and they can submit a bug report. 

> they go on with their lives, but if they get a warning message prompting to submit a bug report with an appropriate log with the backtrace from the server, they are much more likely to report things.

If a completion request fails, then they will see an error message. Is this not enough of an incentive to report the bug? If so, bugs in all other requests are being left unreported.

> So I think it may be better here to show a warning, log the exception, and offer our best completion suggestions we could deliver.

In this particular case, I do not think that "best effort" is a good policy. The benefits are very slim and the additional code is extra weight.

> Thank you for patience for my long arguments

Not at all, I always try to make time for those who contribute to this project. I sense that there's a bit of a difference in maintenance philosophy between us. This project both moves fast and has small dev resources behind it, therefore I'd like it if every line of code would carry its weight. That also means that we are going to leave some edge cases unpolished if the failure mode isn't catastrophic.

Reply:
> If a completion request fails, then they will see an error message.

Just to be on the same page, in vscode, there will be no error message shown, it's just there will be no completion suggestions, that's it. (same situation with all exceptions thrown by ocaml-lsp) The user simply doesn't get completion suggestions and needs to infer that they might need to go to the `Output` pane to see if there is an exception message buried somewhere, assuming (1) they know about the output pane (I'm working on doc "how to debug") and (2) they have verbose mode on. Both assumptions are rare in my opinion. And in the worst case, they get their language server crash 5 times (very unlikely given how stable completion is), at which point they'll need to reload vscode. (this info is more FYI for future than to change your mind on this; I'm okay with an exception, as completion has been quite stable) 

> Not at all, I always try to make time for those who contribute to this project. 

I appreciate it a lot! :-)

Reply:
That's a very good point. I've been meaning to explore a more systematic way of reporting these issues using your metrics PR + [telemetry](https://github.com/microsoft/vscode-extension-telemetry). We should revisit this at some point. In any case, this seems like an issue for all requests (what happens when offering code actions fails?), so we should treat it uniformly.

For now, we can add a workaround that sends a warning hinting that we may need a bug report if a request fails.